crates/integration_test/src/tests/readonly_paths/readonly_paths.rs

```diff
@@ -0,0 +1,275 @@
+    let ro_paths = vec![root.join(&ro_device).to_string_lossy().to_string()];
+
+    let spec = get_spec(ro_paths);
+
+    test_inside_container(spec, &|bundle| {
+        use std::{fs, io};
+
+        let bundle_path = bundle.as_ref();
+        let test_file = bundle_path.join(&ro_device);
+        // NOTE
+        // yes, I know using unsafe willy-nilly is a bad idea,
+        // especially given that OpenOptionsExt in std::os::unix::fs does provide a method
+        // to set mode in open options, like this :
+
+        // use std::os::unix::fs::OpenOptionsExt;
+        // let mut opts = fs::OpenOptions::new();
+        // opts.mode(mode);
+        // opts.create(true);
+        // if let io::Result::Err(e) = opts.open(&test_file) {
+        //     bail!(
+        //         "could not create device node at {:?} with mode {}, got error {:?}",
+        //         test_file,
+        //         mode ^ 0o666,
+        //         e
+        //     );
+        // }
+
+        // but that gives OsErr 22, invalid arguments.
+        // That is why we directly use mknod from lib here
+
+        let _path = test_file.to_string_lossy().as_ptr() as *const i8;
+        let r = unsafe { libc::mknod(_path, mode, 0) };
+        if r != 0 {
+            bail!(
+                "error in creating a device node at {:?} with mode {:?}, got return code {}",
+                test_file,
+                mode,
+                r
+            );
+        }
+
+        match fs::metadata(&test_file) {
+            io::Result::Ok(_) => Ok(()),
+            io::Result::Err(e) => {
+                bail!("error in creating device node, {:?}", e)
+            }
+        }
+    })
+}
+
+fn check_readonly_device_nodes() -> TestResult {
+    let modes = [
+        libc::S_IFBLK | 0o666,
+        libc::S_IFCHR | 0o666,
+        libc::S_IFIFO | 0o666,
+    ];
+    for mode in modes {
+        let res = test_node(mode);
+        if let TestResult::Failed(_) = res {
+            return res;
+        }
+        std::thread::sleep(std::time::Duration::from_millis(1000));
+    }
+    TestResult::Passed
+}
+
+pub fn get_ro_paths_test<'a>() -> TestGroup<'a> {
+    let ro_paths = Test::new("readonly_paths", Box::new(check_readonly_paths));
+    let ro_rel_paths = Test::new("readonly_rel_paths", Box::new(check_readonly_rel_path));
+    let ro_symlinks = Test::new("readonly_symlinks", Box::new(check_readonly_symlinks));
+    // let ro_device_nodes = Test::new(
+    //     "readonly_device_nodes",
+    //     Box::new(check_readonly_device_nodes),
+    // );
+    let ro_device_nodes_blk = Test::new(
+        "readonly_device_nodes_blk",
+        Box::new(|| test_node(libc::S_IFBLK | 0o666)),
+    );
+    let ro_device_nodes_chr = Test::new(
+        "readonly_device_node_chr",
+        Box::new(|| test_node(libc::S_IFCHR | 0o666)),
+    );
+
+    let ro_device_nodes_fifo = Test::new(
+        "readonly_device_nodes_fifo",
+        Box::new(|| test_node(libc::S_IFIFO | 0o666)),
+    );
```

Comment:
Another issue with the device node is that if they are run separately, as they are here then all three tests pass. If I try to put them through a loop, as in the commented test function, then only the first in the loop succeeds, and the second fails. I tried this with different permutations, and always the first iteration of loops succeeds, the second one fails, irrespective device type which is first and second. I suspected that it might be issue that I'm trying to bind same device major & minor (0 in the test_device) to block / char / fifo device node simultaneously, but 
- as tests are run one after other in loop, this should not be an issue
- I tried binding same device (major,minor) = (0,0) locally manually simultaneously to blk/charr/fifo devices using mknod, and that actually succeeds.

I cannot understand what issue might be there, as each device node will be created in a separate container, which separate invocations of check_inside_runtime in the loop version

Reply:
solved by using OpenOptions with `create(true).write(true)`